icu4c/source/i18n/messageformat2_function_registry.cpp

```diff
@@ -905,6 +906,181 @@ Formatter* StandardFunctions::DateTimeFactory::createFormatter(const Locale& loc
     return result;
 }
 
+static UDate tryPattern(const char* pat, const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    LocalPointer<DateFormat> dateParser(new SimpleDateFormat(UnicodeString(pat), errorCode));
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    return dateParser->parse(sourceStr, errorCode);
+}
+
+static UDate tryPatterns(const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    UDate d = tryPattern("YYYY-MM-dd'T'HH:mm:ss", sourceStr, errorCode);
+    if (U_FAILURE(errorCode)) {
+        errorCode = U_ZERO_ERROR;
+        d = tryPattern("YYYY-MM-dd", sourceStr, errorCode);
+    }
+    return d;
+}
+
+static UDate tryTimeZonePatterns(const UnicodeString& sourceStr, UErrorCode& errorCode) {
+    if (U_FAILURE(errorCode)) {
+        return 0;
+    }
+    UDate d = tryPattern("YYYY-MM-dd'T'HH:mm:sszzzz", sourceStr, errorCode);
```

Comment:
instead of trying first one then the next one, could you make decision of which one to call based on the length and  last character in sourceStr? Why are we trying theset two different pattern? Where in the spec specify it take both zzzz and Z ? (how about other pattern?)

Reply:
Done